vignettes/wflow-06-gitlab.Rmd

```diff
@@ -0,0 +1,147 @@
+---
+title: "Hosting worklowr using GitLab"
+subtitle: "workflowr version `r utils::packageVersion('workflowr')`"
+author: "John Blischak"
+date: "`r Sys.Date()`"
+output:
+  rmarkdown::html_vignette:
+    toc: true
+vignette: >
+  %\VignetteIndexEntry{Frequently asked questions}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## What is in this vignette?
+
+By default workflowr assumes that the project will be hosted on GitHub, but this
+is not always the case. Users may prefer to use another service or have a
+private git repository hosting instance. This vignette details some of the steps
+required to host a workflowr project on GitLab. Similar steps will be required
+for other platforms but some of the specifics will be different.
+
+As these steps are not currently integrated into workflowr this vignette assumes
+a basic knowledge of using git from a command line interface.
+
+## Step 0: Set up a project
+
+The first thing we need to do is set up the project we want to host. We can do
+this by following the first few steps of the instructions in the
+"Getting started" vignette, however we can't use the `wflow_git_remote`
+command as it assumes we will be using GitHub. When you get to that stage follow
+the rest of the steps in this vignette.
+
+## Step 1: Set up a remote repository
+
+Log in to the GitLab instance you want to use and create a repository to host
+your project. You will need to know your user name and the repository name for
+the following steps (here we are going to use "myname" and "myproject") as well
+as a URL for the hosting instance.
+
+We now want to connect our local git repository to the one we have just created
+on GitLab. Helpully GitLab provides us instructions on how to do that under
+"Existing Git respository" on our new repository page. The simplest thing to
+do is copy these commands and run them in your local repository directory. If
+you can't see the commands or want to do it manually the most important parts
+are:
+
+```
+git remote add origin HOSTING_URL/myname/myproject.git
```

Comment:
I'd like to update `wflow_git_remote` to be more flexible. Could you please confirm that your custom GitLab instance follows the same pattern as GitHub and GitLab for HTTPS vs. SSH remotes?

GitHub: `https://github.com` vs. `git@github.com`
GitLab: `https://gitlab.com` vs. `git@gitlab.com`
Custom: `https://<domain>` vs. `git@<domain>`

If your custom domain also works like this, then I can add an argument `domain` to `wflow_git_remote`, with the default set to `"github.com"`. This would preserve backwards compatibility as well as provide the extra flexibility needed to support diverse hosting options.

Reply:
Pretty sure it's the same. My git remote is `git@OUR.GIT.DOMAIN:USER/REPO.git`.

Reply:
OK. That's great. If all Git hosting platforms use the same protocol (which I guess makes sense since they are all communicating via Git), then hopefully I should be able to make workflowr agnostic to the Git hosting site (at least for the main features, I'm not sure about links to past files).

Reply:
@pcarbo Could you please confirm that the GitLab instance hosted by RCC Midway also follows this pattern?

Reply:
I think the actual webpage link might be a bit different to GitHub, mine is http://USER.pages.DOMAIN/REPO/

Reply:
@lazappi That actually should not be a problem. workflowr only needs to be able to navigate the directory structure of the repository on the GitLab instance. The actual final URL of the current version of the website isn't used.

Reply:
It is for links isn't it? That's why you need the `github` attribute in `_workflowr.yml`?

Reply:
I might have misunderstood. Is `http://USER.pages.DOMAIN/REPO/` the URL for the Git repository or the URL for the website hosted by GitLab Pages? If it's the URL for the Git repository, then we have a problem. We'd need to include in the documentation that hosting on a custom GitLab instance will require setting `github` in `_workflowr.yml`.

As a contrast, using gitlab.com is just like GitHub, so the user wouldn't need to set `github` (i.e. workflowr can guess based on the remote URL).

URL for Git repository: `https://gitlab.com/USER/REPO`
URL for website: `https://USER.gitlab.io/REPO/`

Reply:
Ah, ok. I think it is slightly different for a self-hosted GitLab which will require setting the `github` attribute.

My URLs are:

Git repository: `http://git.DOMAIN/USER/REPO`
Website: `http://USER.pages.DOMAIN/REPO`

So the website URL has `.pages.` before the domain instead of ending in `.io`.

Reply:
@lazappi OK. Good to know. Then I think we should advise manually specifying the URL of the Git repo for any user that is hosting on a custom GitLab instance. Worst case it is redundant information.